zcash_primitives/src/merkle_tree.rs

```diff
@@ -44,590 +23,307 @@ pub trait HashSer {
     fn write<W: Write>(&self, writer: W) -> io::Result<()>;
 }
 
-impl<T> Hashable for T
-where
-    T: incrementalmerkletree::Hashable + HashSer + Copy,
-{
-    /// Parses a node from the given byte source.
-    fn read<R: Read>(reader: R) -> io::Result<Self> {
-        <Self as HashSer>::read(reader)
+impl HashSer for MerkleHashOrchard {
+    fn read<R: Read>(mut reader: R) -> io::Result<Self>
+    where
+        Self: Sized,
+    {
+        let mut repr = [0u8; 32];
+        reader.read_exact(&mut repr)?;
+        <Option<_>>::from(Self::from_bytes(&repr)).ok_or_else(|| {
+            io::Error::new(
+                io::ErrorKind::InvalidInput,
```

Comment:
```suggestion
                io::ErrorKind::InvalidData,
```
https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.InvalidData